pika/adapters/select_connection.py

```diff
@@ -217,120 +195,202 @@ def add_timeout(self, deadline, callback_method):
+        try:
+            read, write, error = select.select(self._fd_events[READ],
+                                               self._fd_events[WRITE],
+                                               self._fd_events[ERROR],
+                                               self.get_next_deadline())
+        except select.error as error:
+            if error.errno != errno.EINTR:
+                raise
 
-    def remove_timeout(self, timeout_id):
-        """Remove a timeout if it's still in the timeout stack
+        # Build an event bit mask for each fileno we've recieved an event for
 
-        :param str timeout_id: The timeout id to remove
+        fd_event_map = defaultdict(int)
+        for fd_set, ev in zip((read, write, error), (READ, WRITE, ERROR)):
+            for fileno in fd_set:
+                fd_event_map[fileno] |= ev
+
+        self._process_fd_events(fd_event_map, write_only)
+
+    def _process_fd_events(self, fd_event_map, write_only):
+        """ Processes the callbacks for each fileno we've recieved events.
+            Before doing so we re-calculate the event mask based on what is 
+            currently set in case it has been changed under our feet by a 
+            previous callback. We also take a store a refernce to the
+            fd_event_map in the class so that we can detect removal of an
+            fileno during processing of another callback and not generate
+            spurious callbacks on it.
 
+            :param dict fd_event_map: Map of fds to events recieved on them.
         """
-        for i in xrange(len(self._timeouts)):
-            t_id, timeout = self._timeouts[i]
-            if t_id == timeout_id:
-                del self._timeouts[i]
-                break
 
-    def start(self):
-        """Start the main poller loop. It will loop here until self.closed"""
-        while self.open:
-            self.poll()
-            self.process_timeouts()
-            self._manage_event_state()
+        self._processing_fd_event_map = fd_event_map
 
-    def update_handler(self, fileno, events):
-        """Set the events to the current events
+        for fileno in fd_event_map.keys():
+            if fileno not in fd_event_map:
+                # the fileno has been removed from the map under our feet.
+                continue
 
-        :param int fileno: The file descriptor
-        :param int events: The event mask
+            events = fd_event_map[fileno]
+            for ev in [READ, WRITE, ERROR]:
+                if fileno not in self._fd_events[ev]:
+                    events &= ~ev
 
+            if events:
+                handler = self._fd_handlers[fileno]
+                handler(fileno, events, write_only=write_only)
+
+    def flush_outbound(self):
+        """Call the state manager who will figure out that we need to write
+            then call the poller's poll function to force it to process events.
         """
-        self.events = events
+        self._()
```

Comment:
@wjps, this call will crash. Also, this `flush_outbound()` method didn't exist prior to this PR and appears to not be used by anything internally, so should be safe to just delete this method.